File: 프로그래머스/lv3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.py
def distribute(name,proceeds,recommender):
    if name == '-':
        recommender[name][1] += proceeds
        return
    if proceeds*0.1 < 1:
        recommender[name][1] += proceeds
    else:
        distribution = int(proceeds * 0.1)
        recommender[name][1] += proceeds - distribution
        distribute(recommender[name][0],distribution,recommender)
    return

def solution(enroll, referral, seller, amount):
    answer = []
    recommender = {}
    recommender['-'] = ['',0]
    for i in range(len(enroll)):
        recommender[enroll[i]] = [referral[i],0]
    for i in range(len(seller)):
        distribute(seller[i],amount[i]*100,recommender)
    for i in range(len(enroll)):
        answer.append(recommender[enroll[i]][1])
    return answer

File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.java
import java.util.*;

class Solution {

    public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount) {
        int[] answer = new int[enroll.length];
        HashMap<String,String> referralMap = new HashMap<>();
        HashMap<String,Integer> moneyMap = new HashMap<>();
        for(int i=0; i<enroll.length; i++) {
            referralMap.put(enroll[i],referral[i]);
            moneyMap.put(enroll[i],0);
        }

        moneyMap.put("-",0);

        for(int i=0; i< seller.length; i++) {
            dfs(seller[i],amount[i]*100,referralMap,moneyMap);
        }

        for (int i = 0; i < enroll.length; i++) {
            answer[i] = moneyMap.get(enroll[i]);
        }

        return answer;
    }

    public void dfs(String name, Integer sales, HashMap<String,String> referralMap, HashMap<String,Integer> moneyMap) {
        int dividend = sales / 10;
        if (name.equals("-") || dividend == 0) {
            moneyMap.put(name,moneyMap.get(name)+sales);
            return;
        }
        moneyMap.put(name, moneyMap.get(name) + sales - dividend);
        dfs(referralMap.get(name), dividend, referralMap, moneyMap);
    }
}